File: scripts/lib/llm.js
/**
 * llm.js - Battery-included LLM wrapper
 *
 * Usage:
 *   const { ask, chat, stream, askJson, embed } = require('./lib/llm');
 *
 *   // One-liner
 *   const answer = await ask("What is 2+2?");
 *
 *   // With system prompt
 *   const answer = await ask("Summarize this", { system: "You are a summarizer" });
 *
 *   // Override model per-call
 *   const answer = await ask("Hello", { model: "gpt-4o" });
 *
 *   // Multi-turn conversation
 *   const answer = await chat([
 *     { role: "system", content: "You are helpful" },
 *     { role: "user", content: "Hi" },
 *   ]);
 *
 *   // Streaming
 *   for await (const chunk of stream("Tell me a story")) {
 *     process.stdout.write(chunk);
 *   }
 *
 *   // JSON mode
 *   const data = await askJson("List 3 colors as JSON array");
 *
 *   // Embeddings
 *   const vec = await embed("Hello world");
 *
 * Env vars:
 *   OPENAI_API_KEY                    - API key (required)
 *   LLM_BASE_URL   / OPENAI_BASE_URL  - Base URL (optional, for proxies/local)
 *   LLM_MODEL      / OPENAI_MODEL     - Default model (default: gpt-4o-mini)
 *   LLM_TEMPERATURE                   - Default temperature (default: 0.7)
 *   LLM_MAX_TOKENS                    - Default max tokens (default: 4096)
 */

const { OpenAI } = require('openai');
const pRetry = require('p-retry'); // pinned to v4 (CommonJS)

// --------------- Config from env ---------------

const API_KEY = process.env.OPENAI_API_KEY;
const BASE_URL = process.env.LLM_BASE_URL || process.env.OPENAI_BASE_URL || undefined;
const DEFAULT_MODEL = process.env.LLM_MODEL || process.env.OPENAI_MODEL || 'gpt-4o-mini';
const DEFAULT_TEMP = parseFloat(process.env.LLM_TEMPERATURE || '0.7');
const DEFAULT_MAX_TOKENS = parseInt(process.env.LLM_MAX_TOKENS || '4096', 10);

// --------------- Client (singleton) ---------------

let _client = null;

function getClient() {
  if (!_client) {
    const opts = {};
    if (API_KEY) opts.apiKey = API_KEY;
    if (BASE_URL) opts.baseURL = BASE_URL;
    _client = new OpenAI(opts);
  }
  return _client;
}

function resetClient() {
  _client = null;
}

// --------------- Retry config ---------------

const RETRY_OPTS = {
  retries: 3,
  minTimeout: 2000,
  factor: 2,
  onFailedAttempt: (err) => {
    // Only retry on 429 (rate limit) and 5xx (server errors)
    const status = err?.status || err?.response?.status;
    if (status && status !== 429 && status < 500) {
      throw err; // Don't retry client errors (4xx except 429)
    }
  },
};

// --------------- Core functions ---------------

/**
 * Ask a single question, get a string answer.
 */
async function ask(prompt, opts = {}) {
  const { system, model, temperature, maxTokens, ...rest } = opts;
  const messages = [];
  if (system) messages.push({ role: 'system', content: system });
  messages.push({ role: 'user', content: prompt });
  return chat(messages, { model, temperature, maxTokens, ...rest });
}

/**
 * Send a full conversation, get a string answer.
 */
async function chat(messages, opts = {}) {
  const { model, temperature, maxTokens, ...rest } = opts;
  return pRetry(async () => {
    const resp = await getClient().chat.completions.create({
      model: model || DEFAULT_MODEL,
      messages,
      temperature: temperature ?? DEFAULT_TEMP,
      max_tokens: maxTokens || DEFAULT_MAX_TOKENS,
      ...rest,
    });
    return resp.choices[0].message.content;
  }, RETRY_OPTS);
}

/**
 * Ask and parse the response as JSON.
 */
async function askJson(prompt, opts = {}) {
  const { system, model, ...rest } = opts;
  const sysMsg = ((system || '') + '\nAlways respond with valid JSON.').trim();
  const messages = [
    { role: 'system', content: sysMsg },
    { role: 'user', content: prompt },
  ];
  return pRetry(async () => {
    const resp = await getClient().chat.completions.create({
      model: model || DEFAULT_MODEL,
      messages,
      response_format: { type: 'json_object' },
      temperature: rest.temperature ?? DEFAULT_TEMP,
      max_tokens: rest.maxTokens || DEFAULT_MAX_TOKENS,
    });
    return JSON.parse(resp.choices[0].message.content);
  }, RETRY_OPTS);
}

/**
 * Stream a response token-by-token (async generator).
 */
async function* stream(prompt, opts = {}) {
  const { system, model, temperature, maxTokens, ...rest } = opts;
  const messages = [];
  if (system) messages.push({ role: 'system', content: system });
  messages.push({ role: 'user', content: prompt });

  const resp = await getClient().chat.completions.create({
    model: model || DEFAULT_MODEL,
    messages,
    temperature: temperature ?? DEFAULT_TEMP,
    max_tokens: maxTokens || DEFAULT_MAX_TOKENS,
    stream: true,
    ...rest,
  });

  for await (const chunk of resp) {
    const delta = chunk.choices[0]?.delta?.content;
    if (delta) yield delta;
  }
}

/**
 * Get embedding vector(s).
 */
async function embed(text, opts = {}) {
  const { model = 'text-embedding-3-small' } = opts;
  const input = Array.isArray(text) ? text : [text];
  return pRetry(async () => {
    const resp = await getClient().embeddings.create({ model, input });
    const vectors = resp.data.map((d) => d.embedding);
    return Array.isArray(text) ? vectors : vectors[0];
  }, RETRY_OPTS);
}

module.exports = {
  ask,
  chat,
  stream,
  askJson,
  embed,
  getClient,
  resetClient,
  DEFAULT_MODEL,
};

File: scripts/lib/llm.py
"""
llm.py - Battery-included LLM wrapper

Usage:
    from lib.llm import ask, chat, stream, embed

    # One-liner
    answer = ask("What is 2+2?")

    # With system prompt
    answer = ask("Summarize this", system="You are a summarizer")

    # Override model per-call
    answer = ask("Hello", model="gpt-4o")

    # Multi-turn conversation
    answer = chat([
        {"role": "system", "content": "You are helpful"},
        {"role": "user", "content": "Hi"},
    ])

    # Streaming
    for chunk in stream("Tell me a story"):
        print(chunk, end="", flush=True)

    # JSON mode
    data = ask_json("List 3 colors as a JSON array")

    # Embeddings
    vec = embed("Hello world")

Env vars:
    OPENAI_API_KEY                    - API key (required)
    LLM_BASE_URL   / OPENAI_BASE_URL  - Base URL (optional, for proxies/local)
    LLM_MODEL                         - Default model (default: gpt-4o-mini)
    LLM_TEMPERATURE                   - Default temperature (default: 0.7)
    LLM_MAX_TOKENS                    - Default max tokens (default: 4096)
"""

import os
import json
from typing import Optional, Generator
from openai import OpenAI
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception

# --------------- Config from env ---------------

API_KEY = os.getenv("OPENAI_API_KEY")
BASE_URL = os.getenv("LLM_BASE_URL") or os.getenv("OPENAI_BASE_URL") or None
DEFAULT_MODEL = os.getenv("LLM_MODEL") or "gpt-4o-mini"
DEFAULT_TEMP = float(os.getenv("LLM_TEMPERATURE", "0.7"))
DEFAULT_MAX_TOKENS = int(os.getenv("LLM_MAX_TOKENS", "4096"))

# --------------- Client (singleton) ---------------

_client: Optional[OpenAI] = None


def get_client() -> OpenAI:
    global _client
    if _client is None:
        kwargs = {}
        if API_KEY:
            kwargs["api_key"] = API_KEY
        if BASE_URL:
            kwargs["base_url"] = BASE_URL
        _client = OpenAI(**kwargs)
    return _client


def reset_client():
    """Force re-create client (e.g. after changing env vars at runtime)."""
    global _client
    _client = None


# --------------- Retry logic ---------------

def _is_retryable(exc):
    """Retry on rate limits (429) and server errors (5xx)."""
    from openai import RateLimitError, APIStatusError
    if isinstance(exc, RateLimitError):
        return True
    if isinstance(exc, APIStatusError) and exc.status_code >= 500:
        return True
    return False


_retry_decorator = retry(
    retry=retry_if_exception(_is_retryable),
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=2, max=30),
    reraise=True,
)


# --------------- Core functions ---------------

@_retry_decorator
def ask(
    prompt: str,
    *,
    system: Optional[str] = None,
    model: Optional[str] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
    **kwargs,
) -> str:
    """Ask a single question, get a string answer."""
    messages = []
    if system:
        messages.append({"role": "system", "content": system})
    messages.append({"role": "user", "content": prompt})
    return chat(messages, model=model, temperature=temperature, max_tokens=max_tokens, **kwargs)


@_retry_decorator
def chat(
    messages: list,
    *,
    model: Optional[str] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
    **kwargs,
) -> str:
    """Send a full conversation, get a string answer."""
    resp = get_client().chat.completions.create(
        model=model or DEFAULT_MODEL,
        messages=messages,
        temperature=temperature if temperature is not None else DEFAULT_TEMP,
        max_tokens=max_tokens or DEFAULT_MAX_TOKENS,
        **kwargs,
    )
    return resp.choices[0].message.content


@_retry_decorator
def ask_json(
    prompt: str,
    *,
    system: Optional[str] = None,
    model: Optional[str] = None,
    **kwargs,
) -> dict | list:
    """Ask and parse the response as JSON. Uses json_object response format."""
    sys_msg = (system or "") + "\nAlways respond with valid JSON."
    messages = [
        {"role": "system", "content": sys_msg.strip()},
        {"role": "user", "content": prompt},
    ]
    resp = get_client().chat.completions.create(
        model=model or DEFAULT_MODEL,
        messages=messages,
        response_format={"type": "json_object"},
        temperature=kwargs.pop("temperature", DEFAULT_TEMP),
        max_tokens=kwargs.pop("max_tokens", DEFAULT_MAX_TOKENS),
        **kwargs,
    )
    return json.loads(resp.choices[0].message.content)


def stream(
    prompt: str,
    *,
    system: Optional[str] = None,
    model: Optional[str] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
    **kwargs,
) -> Generator[str, None, None]:
    """Stream a response token-by-token."""
    messages = []
    if system:
        messages.append({"role": "system", "content": system})
    messages.append({"role": "user", "content": prompt})

    resp = get_client().chat.completions.create(
        model=model or DEFAULT_MODEL,
        messages=messages,
        temperature=temperature if temperature is not None else DEFAULT_TEMP,
        max_tokens=max_tokens or DEFAULT_MAX_TOKENS,
        stream=True,
        **kwargs,
    )
    for chunk in resp:
        delta = chunk.choices[0].delta.content
        if delta:
            yield delta


@_retry_decorator
def embed(
    text: str | list[str],
    *,
    model: str = "text-embedding-3-small",
) -> list:
    """Get embedding vector(s). Returns single list for string, list of lists for list input."""
    inp = text if isinstance(text, list) else [text]
    resp = get_client().embeddings.create(model=model, input=inp)
    vectors = [d.embedding for d in resp.data]
    return vectors if isinstance(text, list) else vectors[0]
